gdal/alg/gdalwarpkernel.cpp

```diff
@@ -5821,49 +5831,39 @@ static void GWKAverageOrModeThread( void* pData)
         }
         else
         {
-            nAlgo = GWKAOM_Fmode;
+            nAlgo = AOM::FloatMode;
 
-            if( nSrcXSize > 0 && nSrcYSize > 0 )
-            {
-                pafRealVals = static_cast<float *>(
-                    VSI_MALLOC3_VERBOSE(nSrcXSize, nSrcYSize, sizeof(float)));
-                panRealSums = static_cast<int *>(
-                    VSI_MALLOC3_VERBOSE(nSrcXSize, nSrcYSize, sizeof(int)));
-                if( pafRealVals == nullptr || panRealSums == nullptr )
-                {
-                    VSIFree(pafRealVals);
-                    VSIFree(panRealSums);
-                    return;
-                }
-            }
+            // Cast to size_t to prevent overflow.
+            realVals.resize(static_cast<size_t>(nSrcXSize) * nSrcYSize);
```

Comment:
could you also put that in a try {} catch {} block, with a CPLError(CE_Failure, CPLE_OutOfMemory, "%s", e.what()) in the catch {},  if resize() fails due to out-of-memory ?

Reply:
What do you think about a function that returns a vector for this purpose? Something like `makeRaster()`. Then you don't need to have a local try/catch and you can still avoid the malloc/free mess. And if we decide that the behavior should change, we can do it in one place? That's kind of what `VSIMalloc2()` is doing, but you still have the lingering cleanup to deal with.

Reply:
> What do you think about a function that returns a vector for this purpose? Something like `makeRaster()`

How would that behave in OOM situations ? I guess I could better comment on actual code :-)

Reply:
Updated with a sample function.

Reply:
> Updated with a sample function.

ok, I see. Why not. But having exceptions thrown here will cause issues in the call stack. They have to be caught at some level, and we need to make sure all code between where they are caught and thrown properly cleans up after itself. The warp kernels can be invoked either in single thread usage or from a worker thead.

Reply:
I still believe that if this situation occurs, recovery isn't really a good option.  IMO, the chances that you actually get a result that you're going to be pleased with approaches zero.

That said, `bad_alloc` can happen at any time. You can do some string operation and get a bad_alloc. If we're concerned about threads terminating "cleanly" (and I'm still not sure that matters), you can do that at the thread pool level or similar. There are all kinds of exceptions that can be thrown when you're using C++11, so if this is a concern, it seems a general one and not specific to this case. Handling the exceptions at the thread main or application main seems reasonable if necessary, though I'm not quite sure what you'd do other than discard, since we've already provided some user feedback.

The current handling depends on every caller up the stack noticing that there's a NULL returned and figuring out what to do about it, which seems much more likely to end in unexpected handling than if it were more centralized, but I guess that's also what I'm used to.

Reply:
> I still believe that if this situation occurs, recovery isn't really a good option. IMO, the chances that you actually get a result that you're going to be pleased with approaches zero.

Propagating the error to the caller in one way or another is what we generally do when VSIMalloc() fails. There's indeed no way to 'recover' / continue the processing.

> That said, bad_alloc can happen at any time. 

Agreed we're not completely bullet proof currently, but the likelihood that a bad_alloc occurs for big alloc like that one is much higher than for small string manipulations.

My whole point is that I want our error handling to be at least as good/bad as it is currently. Understanding we could be more bullet proof but that will require extensive and progressive changes through the code base.

Reply:
> Agreed we're not completely bullet proof currently, but the likelihood that a bad_alloc occurs for big alloc like that one is much higher than for small string manipulations.

Of course. The example code should catch that event. The ultimate question is: "what now?"

> My whole point is that I want our error handling to be at least as good/bad as it is currently. Understanding we could be more bullet proof but that will require extensive and progressive changes through the code base.

This is why I made RFC 82. But the nice thing about exceptions is that we _don't_ have to fix all the code. We can simply change the thread pool code of the application main to deal with it as we like. I'm just not sure what that looks like.

For the record, I'm not an exception fan in general, but for something like this, they seem a much better solution than what's currently happening. And more important to me, it allows for cleaner code that's less likely to leak.

Reply:
> But the nice thing about exceptions is that we _don't_ have to fix all the code.

We can create leaks that didn't exist before. That's my concern.

if you have:
```
bool foo()
{
   char* x = malloc(1);
   bool ret = bar();
   if( ret )
   { 
        ...
   }
   free(x);
   return ret;
}

bool bar()
{
     void* big_alloc = malloc( 1 GB );
     if( !big_alloc )
           return false;
     ...
     free(big_alloc);
     return true;
}
```
and you replace the malloc() by a new[], then you'll create a leak.

Reply:
I don't understand this comment. I don't see how it's applicable to this PR at all. I'm hoping to take advantage of RAII, not just replace `malloc` with `new`, which seems pointless to me.

Reply:
> not just replace `malloc` with `new`

sorry, I mean using resize(size), or std::vector<>(initial_size), but my above point still holds.  Having bad_alloc emitted for callers not ready for it can cause memory leaks.

Reply:
You're concerned about memory leaks caused by old-style malloc/free handling in the case where you throw an exception because you're out of heap that bypasses a `free`? That seems like a strange worry to me, but I won't argue it with you.

The GDAL code base is too large to address this kind of thing all at once. It'll never happen. So you either have to accept this and make progress over time or never do anything.

Reply:
> So you either have to accept this and make progress over time or never do anything.

if we wanted to throw exceptions, yes, we'd have to do this carefully and progressively, and review all the callers to make sure they use RAII patterns.
Addressing more RFC 81 than this pull request: but I'm not sure we want to propagate exceptions at least at the C++ API level. Let's say for example GDALDataset::RasterIO() which could for example go to through the warping code if using a warped VRT. If we wanted to do that, we'd have to advertize this might happen now (currently it should happen very rarely), and more importantly edit all C API entry points so it catches exceptions.
Actually the logical sequence would be:
1. catch exceptions in C API entry points
2. advertise the C++ API might now throw exceptions for users of the C++ API
3. progressive edit the code to use RAII patterns and potentially use std::vector for 'big' allocs that can cause exceptions to be thrown.

Reply:
I've tried to be clear here, but I don't think any progress is being made.

When you're out of memory, it's a fatal condition. Worrying about an exception being thrown is silly and worrying about people catching such a condition (that they might *currently* have to expect) is just foot-dragging, IMO.

I don't really have any interest in working on software that can't be improved in a meaningful way, so I won't. Maybe others will be able to make some headway.